### PR TITLE
Shunt incremental voltage control

### PR DESCRIPTION
**Pull request by Hadrien-Godard**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
This PR introduces a new mode for shunt voltage control. This new mode is named shunt incremental voltage control and allows shunt controlers to increment tap per tap their tap position to reach their target voltage deadband. Therefore, this mode takes controler deadbands into account, which was not the case in the previous implemented mode.

**What is the current behavior?** *(You can also link to an open issue here)*
Current behavior is that there is only one mode for shunt voltage control. This mode does not increment tap positions. This mode solves the load flow equation where the shunt susceptance is a continuous variable and then round the susceptance to the closest tap position.

**What is the new behavior (if this is a feature change)?**
With the new incremental mode, tap positions are incremented for all controlers in each outer loop iteration, converging when all controlers reach their deadband.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Comment by AnkurArohi** [comment]
There is no proper description of what this pr is about please do fill the questionnaire

File: src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+/**
+ * @author Hadrien Godard <hadrien.godard at artelys.com>
+ */
+
+public class IncrementalShuntVoltageControlOuterLoop implements OuterLoop {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(IncrementalShuntVoltageControlOuterLoop.class);
+
+    private static final int MAX_DIRECTION_CHANGE = 2;
```
Comment (AnkurArohi) [inline src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java:37]: Please comment what is meant by this parameter

File: src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        }
+    }
+
+    private void adjustB(List<LfShunt> sortedControllerShunts, LfBus controlledBus, ContextData contextData, int[] controllerShuntIndex,
+                                              DenseMatrix sensitivities, double diffV, MutableObject<OuterLoopStatus> status) {
+        // several shunts control the same bus
+        double remainingDiffV = diffV;
+        boolean hasChanged = true;
+        while (hasChanged) {
+            hasChanged = false;
```
Comment (AnkurArohi) [inline src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java:115]: This is not a conform code, there has to be a better way to do it

File: src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                    if (checkTargetDeadband(targetDeadband, remainingDiffV)) {
+                        double previousB = controller.getB();
+                        double deltaB = remainingDiffV / sensitivity;
+                        LfShuntImpl.Controller.Direction direction = controller.updateTapPositionB(deltaB, 1, controllerContext.getAllowedDirection()).orElse(null);
+                        if (direction != null) {
+                            updateAllowedDirection(controllerContext, direction);
+                            remainingDiffV -= (controller.getB() - previousB) * sensitivity;
+                            hasChanged = true;
+                            status.setValue(OuterLoopStatus.UNSTABLE);
+                            LOGGER.debug("Increment shunt susceptance of '{}': {} -> {}", controller.getId(), previousB, controller.getB());
```
Comment (AnkurArohi) [inline src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java:135]: Susceptence value

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 4]
```diff
@@ -155,8 +158,15 @@ public enum TransformerVoltageControlMode {
         INCREMENTAL_VOLTAGE_CONTROL
     }
 
+    public enum ShuntVoltageControlMode {
+        WITH_GENERATOR_VOLTAGE_CONTROL,
+        INCREMENTAL_VOLTAGE_CONTROL
+    }
+
     public static final TransformerVoltageControlMode TRANSFORMER_VOLTAGE_CONTROL_MODE_DEFAULT_VALUE = TransformerVoltageControlMode.WITH_GENERATOR_VOLTAGE_CONTROL;
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:166]: Both default values of voltage control for transformers and for shunts can be changed in the future, if the incremental outerloops are mature enough.

File: src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java (hunk 3 above)
Comment (Hadrien-Godard) [reply src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java:135]: Just to check, this comment means that it is more proper to write  "Increment shunt susceptance value of" in the log?

Comment (AnkurArohi) [reply src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java:135]: yes

File: src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        List<LfShunt> controllerShunts = getControllerShunts(network);
+        SensitivityContext sensitivityContext = new SensitivityContext(network, controllerShunts,
+                loadFlowContext.getEquationSystem(), loadFlowContext.getJacobianMatrix());
+
+        network.getBuses().stream()
+                .filter(LfBus::isShuntVoltageControlled)
+                .forEach(controlledBus -> {
+                    ShuntVoltageControl voltageControl = controlledBus.getShuntVoltageControl().orElseThrow();
+                    double diffV = voltageControl.getTargetValue() - voltageControl.getControlled().getV();
+                    List<LfShunt> sortedControllers = voltageControl.getControllers().stream().sorted(Comparator.comparing(LfShunt::getBMagnitude)).collect(Collectors.toList());
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/IncrementalShuntVoltageControlOuterLoop.java:208]: @Hadrien-Godard do we need a `.reversed()` here? What do you want?

**Review by annetill — APPROVED** [review]

**Review by geofjamg — APPROVED** [review]

**Pull request merged**